students/Cracolici_Jon/lesson05/JonCracolici_Mailroom3.py

```diff
@@ -0,0 +1,180 @@
+    """Creates a summary report for all current donors."""
+    dictlist = OrderedDict(sorted(dB.items(), key=lambda x: x[1]['totaldonated'], reverse=True))
+    print('Donor Name                | Total Given | Num Gifts | Average Gift')
+    print('------------------------------------------------------------------')
+    for i in dictlist:
+        print('{donor:26}${totaldonated:>13.2f}{numdonations:>12} ${avgdonated:>12.2f}'.format(**dB[i]))
+    return
+def mass_mail(dB):
+    for item in dB.values():
+        print_letters(item)
+    return start_control(dB)
+def nd_control(dB):
+    """Function that calls functions to create and add a new donor to database, print letter to screen."""
+    newdonordict = donor_input(dB)
+    if newdonordict == False:
+        return False
+    elif newdonordict['donornamekey'] in dB.keys():
+        print("It appears that donor is already in our system.\n Perhaps you would like to select the 'Update Donor' option instead?")
+        return False
+    #will add try except to check for donor actually being new in L5
+    new_donor(newdonordict, dB)
+    note_gen(dB[newdonordict['donornamekey']])
+    return dB#, start_control(dB)
+def ud_control(dB):
+    """Function that calls functions to update a donor in the database, print letter to screen."""
+    updonordict = donor_input(dB)
+    if updonordict == False:
+        return False
+    elif updonordict['donornamekey'] not in dB.keys():
+        print("It appears that you are attempting t update a donor not already in our system.\n Perhaps you would like to select the 'New Donor' option instead?")
+        return False
+    #will add try except to check for donor actually being new in L5
+    update_donor(updonordict, dB)
+    note_gen(dB[updonordict['donornamekey']])
+    return dB#, start_control(dB)
+def donor_input(dB):
+    """This function collects user input about the donation."""
+    #create a set of quit criteria
+    quitset = set(['quit', 'q'])
+    donorname = input('Who is the donor?')
+    #Check if user wants to exit to previous menu
+    if donorname.lower() in quitset:
+        return False
+    newdonation = input('How much did they donate?')
+    #Check if user wants to exit to previous menu
+    if newdonation.lower() in quitset:
+        return False
+        
+    else:#this section creates the keys and values for the outputted dict
+        donornamekey = donorname.lower()
+        donornamekey = donornamekey.replace(' ', '_')
+        donor = donorname.title()
+        try:
+            donation = float(newdonation)
+        except ValueError:
+            print("Use numbers for the donation amount. NUMBERS!")
+            return False
+        newdondict = {'donornamekey': donornamekey, 'donor': donor, 'donation': donation}
+        return newdondict
+def print_letters(donordict):
+    """This function takes a string and makes a document with it."""
+    lettername = donordict['fname']+'.txt'
+    outfile = open(lettername, 'w')
```

Comment:
prefer to use `with` statement when working with files

Reply:
I have heard that it is preferred because it automatically closes the file. but don't people just close the files regardless? Or is there some other reason?

Reply:
Exactly, it ensures the file is closed in case of an exception, and the code is cleaner.

Not using with is considered an anti-pattern:
https://docs.quantifiedcode.com/python-anti-patterns/maintainability/not_using_with_to_open_files.html

Reply:
that website is pretty cool. I will switch to "with"

Reply:
yes it is! recommend reading more of their posts.